mlflow/onnx.py

```diff
@@ -182,10 +182,27 @@ def _load_model(model_file):
 
 
 class _OnnxModelWrapper:
-    def __init__(self, path):
+    def __init__(self, path, providers=None):
         import onnxruntime
 
-        self.rt = onnxruntime.InferenceSession(path)
+        if providers is None:
+            providers = ['CUDAExecutionProvider', 'CPUExecutionProvider']
+
+        # NOTE: Some distributions of onnxruntime require the specification of the providers argument on calling. E.g. onnxruntime-gpu.
+        #       The package import call does not differnetiate which architecture specific version has been installed, as all are imported with onnxruntime.
+        #       onnxruntime documentation says that from v1.9.0 some distributions require the providers list to be provided on calling an InferenceSession.
+        #       Therefore the try catch nested structure below attempts to create an inference session with just the model path as pre v1.9.0.
+        #       If that fails, it will use the providers list call. 
+        #       At the moment this is just CUDA and CPU, and probably should be expanded. It's been added to the class init call as an optional argument for further user customisation.
+        try:
+            self.rt = onnxruntime.InferenceSession(path)
+        except:
```

Comment:
Could you catch the appropriate exception here to initiate the alternate processing logic?

Reply:
So, if I understand you correctly, catch the exception explicitly by type and then call the alternative loading, instead of a nested try implementation that I have done? Did I understand you right?

In the case where the `providers` variable is needed, the exception thrown is a `ValueError`.

So the refactored version would look as follows:

```python
try:
   self.rt = onnxruntime.InferenceSession(path)
except ValueError:
   self.rt = onnxruntime.InferenceSession(path, providers=providers)
except:
   raise
```

Reply:
exactly! Just so that we can make sure we're only trying the other logic in the case of the exact type of error that we're expecting (Sorry, I was hoping that the onnx package had a custom exception other than the blanket ValueError, but it's generally a good practice to catch only the Exceptions that we want to catch to do any further processing).

Reply:
@BenWilson2 @dbczumar 

Refactor committed with the suggested changes catching the ValueError and then running the post v1.9.0 call with providers.

Added functionality in the `__init__` method of the `_OnnxModelWrapper` class. 
This now reads the `MLModel` file from the same directory as the `model.onnx` has been downloaded to. **The assumption is that the MLModel file will always be present, as the model will be logged using the onnx flavour**.
It attempts to find the field "providers" in the onnx model flavour fields of the MLModel yaml.
If does not find it (for backwards compatability), it uses the default list which is hardwired as an attribute `ONNX_EXECUTION_PROVIDES` of the onnx model flavour.